doc/packaging.md

```diff
@@ -315,9 +315,9 @@ Notes on behavior:
 
 - The imported device CA will now be referred to as the "Edge CA".
 
-- The master identity key and master encryption key are two symmetric keys dynamically generated by `iotedged` for internal use. They will not be imported into the new services.
+- The workload API exposes an encrypt and decrypt API for modules which uses module-specific keys derived from the module ID and a master encryption key. The master encryption key will be imported into the new services.
 
-  Note that this means Edge Agent, Edge Hub and other modules will not be able to decrypt any data when running against the new services that they previously encrypted using the workload API with the old service. This is not a problem for Edge Agent and Edge Hub, in the sense that they ignore the undecryptable data and start with empty storage, so they recover at the cost of data loss.
+- The format of encrypted secrets is changing between v1.1 and v1.2. Using the imported master encryption key, modules that use the workload API and persist across the upgrade from v1.1 -> v1.2 can read secrets encrypted in v1.1. However, upon re-encrypting a secret it is saved in the new format. This means that on a downgrade from v1.2 to v1.1 those modules must be re-deployed. They will be unable to decrypt secrets in the new format. Be aware that in all versions of IoT Edge a module cannot decrypt data previously encrypted by a module of the same name in a previous deployment (i.e. secrets cannot be shared across module generations).
```

Comment:
>Be aware that in all versions of IoT Edge a module cannot decrypt data previously encrypted by a module of the same name in a previous deployment (i.e. secrets cannot be shared across module generations).

This is not true. Generation ID in the encrypt / decrypt request is controlled by the module. If a module can know a previous life's generation ID (say it persists the generation ID unencrypted alongside the ciphertext), it can decrypt the ciphertext in this life.

Reply:
(Similarly, it can also opt to always use a fixed constant for the generation ID and ignore the IoT Hub-set generation ID entirely.)

Reply:
As a general practice would you recommend that customers do either of those (persist the generation ID or use a fixed constant)?  Or would you consider doing that an anti-pattern?

Reply:
Being deployed with a new generation ID kinda implies the module ought to treat it the same as a fresh install, ie it ought to ignore any previous secrets. So on that basis the module ought to use the real generation ID. I can't think of a reason a module would want to retain secrets across deployments.

But who knows what users want. We've given them the power to use an arbitrary generation ID, so maybe some of them (ab)use it.

Reply:
Perhaps something like:

>Be aware that, in all versions of IoT Edge, the key used to encrypt the secrets incorporates the module's name and the generation ID it provides to the encrypt/decrypt API. Thus the module needs to use the same name and generation ID to decrypt a secret that it previously used to encrypt it. If the module uses a new generation ID after it's been re-deployed, it will be unable to decrypt the secrets from a previous deployment.

? That is technically correct while not outright saying that a module can fake its generation ID.

Reply:
Where possible I think we should be prescriptive.  So, in the statement above I'd suggest we include a disclaimer or something to discourage persisting the generation ID.  If we're wrong and its important to customers then we'll learn about it.  And we can always relax the constraint in the future if needed.  So, perhaps...

> after it's been re-deployed, it will be unable to decrypt the secrets from a previous deployment. **We recommend always using the real generation ID.  Persisting a generation ID across deployments to use with the encrypt/decrypt is discouraged.**"

Also, my thought behind the original language was that the individual operating the device and deploying modules isn't necessarily the person who authored the module.  They may be unaware of how it will handle an upgrade/downgrade.  So, as a default recommendation for a downgrade they should be prepared to re-deploy the modules.

Reply:
Okay.

Reply:
The wording here could be a bit confusing - the gen id changes only if the module was removed in an intermediate deployment. If the module is upgraded/changed in a subsequent deployment, then the genId won't change and subsequently the new module should be able to decrypt the secrets left back by the old module.

Reply:
I'm going to move the discussion on encrypting/decrypting into its own section that follows this one on the behavior of `iotedge config import`.